### PR TITLE
Each schema validation

### PR DESCRIPTION
**Pull request by mbenson**
refactor NunjucksWorkflowTaskRunner to:
     - generate minimally informative task log per iteration
     - properly validate iterated actions

Addresses  https://github.com/backstage/backstage/issues/20749

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by benjdlambert** [comment]
@dagda1 @acierto do you wanna take a look at this too and see what you think?

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts [hunk 1]
```diff
@@ -277,73 +277,73 @@ export class NunjucksWorkflowRunner implements WorkflowRunner {
...
-          actionInput =
-            (step.input &&
-              this.render(
-                step.input,
-                { ...iterationContext, secrets: task.secrets ?? {} },
-                renderTemplate,
-              )) ??
-            {};
+        if (iteration.each) {
+          taskLogger.info(
```
Comment (acierto) [inline plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:337]: Can you please add also a unit test for logging?

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts [hunk 2]
```diff
@@ -277,73 +277,73 @@ export class NunjucksWorkflowRunner implements WorkflowRunner {
...
+            `Running step each: ${JSON.stringify(
+              iteration.each,
+              (k, v) => (k ? v.toString() : v),
```
Comment (acierto) [inline plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:340]: I wonder if this is intended, as stringify will do `.toString` for `v`.

**Comment by acierto** [comment]
Looks good to me

**Review by acierto — APPROVED** [review]

**Review by benjdlambert — COMMENTED** [review]
Just a quick comment about the casting.

Also can you explain what you mean by: `- properly validate iterated actions`?

**Comment by mbenson** [comment]
> Just a quick comment about the casting.

What's the comment? I did notice an inconsistency with the way I was checking `iteration.each` but I didn't want to destroy my approval 😛

> Also can you explain what you mean by: - properly validate iterated actions?

Sure:
- Prior to `each`, an action was validated (properly) against any available `input` schema.
- With the initial implementation of `each`, no change was made to the schema validation logic. Unfortunately, the validation was being done against the rendered `input` object without an available `each` property against which to generate any associated input values. Therefore an expression referencing `each` would render to `''`, and if such an expression made up the entirety of a required input value, the schema validation would fail. Thus `each` validation, "improperly." The first commit here is to add a test case that demonstrates this bad behavior when run without the subsequent fix.
- The schema-validation portion of the submitted changes instead runs the validation against each complete set of input data, including the `each` value associated with a given iteration. The existing "allowed" check has been similarly refactored. If any iteration violates either check, the assumption is that the whole set is "bad" and none will therefore be executed.

**Comment by mbenson** [comment]
@benjdlambert to be clear, if you left a comment regarding casting, I couldn't and still can't find it. Also, please see my previous comment for an attempt to clarify the problem that this pull request addresses.

**Comment by benjdlambert** [comment]
Sorry for the delay in getting back here, we're at BackstageCon / KubeCon in Chicago this week so it's been a bit slammed.

Comment about casting was here on this line. It is necessary?: https://github.com/backstage/backstage/pull/20750/files#diff-30eabd978c6d9838c6e26de4af50fce58cc70f1a75b1b577e27b19239db9109aR346

**Comment by mbenson** [comment]
> Sorry for the delay in getting back here, we're at BackstageCon / KubeCon in Chicago this week so it's been a bit slammed.

Ah, makes perfect sense. I had forgotten the convention/s was/were this week! Sorry for nagging under the circumstances.

> 
> Comment about casting was here on this line. It is necessary?: https://github.com/backstage/backstage/pull/20750/files#diff-30eabd978c6d9838c6e26de4af50fce58cc70f1a75b1b577e27b19239db9109aR346
> 

`iterations` was originally defined as `Array<JsonValue>`; I changed that to `Array<JsonObject>` because it was clear we could narrow to that degree of specificity; however that doesn't do anything to enforce the type safety of the values we add to a given (iteration) element, as my refactor does with the `input` associated with a given iteration (this seemed reasonable since that same computed value is consumed in three different ways). We could of course take the refactor a step further by more strongly typing an iteration e.g.:

```
type Iteration = {
  each?: {
    key: any;
    value: JsonValue;
  };
  input?: JsonObject;
};
```

> And thanks for the clarification! 🙏

YW :)

**Comment by mbenson** [comment]
After sleeping on the change I made last night, I was motivated to experiment a bit further, and found that TypeScript seemed to happily infer the `Iteration` type when defined as `const` the way I had done, so I was able to completely eliminate the explicit type declaration, while also removing the original cast you asked about. I have added the final commit to this branch/PR.

**Review by benjdlambert — APPROVED** [review]
Nice thanks for this! :tada:

**Pull request merged**